internal/mq/msgdispatcher/dispatcher.go

```diff
@@ -0,0 +1,263 @@
+		d.done <- struct{}{}
+		d.wg.Wait()
+	case resume:
+		d.wg.Add(1)
+		go d.work()
+	case terminate:
+		d.done <- struct{}{}
+		d.wg.Wait()
+		d.once.Do(func() {
+			d.stream.Close()
+		})
+	default:
+		err := fmt.Errorf("invalid signal in dispatcher handler, pchannel = %s", d.pchannel)
+		log.Error(err.Error())
+		panic(err)
+	}
+	log.Info("dispatcher handled signal done",
+		zap.String("pchannel", d.pchannel),
+		zap.String("signal", signal.string()))
+}
+
+func (d *dispatcher) work() {
+	log.Info("dispatcher begin to work", zap.String("pchannel", d.pchannel))
+	defer d.wg.Done()
+	for {
+		select {
+		case <-d.done:
+			log.Info("dispatcher stopped working", zap.String("pchannel", d.pchannel))
+			return
+		case pack := <-d.stream.Chan():
+			if pack == nil || len(pack.EndPositions) != 1 {
+				log.Error("dispatcher consumed invalid msgPack", zap.String("pchannel", d.pchannel))
+				continue
+			}
+			d.setCurPosition(pack.EndPositions[0])
+
+			// init packs for all target vchannels, even though there's no msg in pack,
+			// but we still need to dispatch time ticks to the targets.
+			d.targetsMu.RLock()
+			packs := make(map[string]*msgstream.MsgPack, len(d.targets))
+			for vchannel := range d.targets {
+				packs[vchannel] = &msgstream.MsgPack{
+					BeginTs:        pack.BeginTs,
+					EndTs:          pack.EndTs,
+					Msgs:           make([]msgstream.TsMsg, 0),
+					StartPositions: pack.StartPositions,
+					EndPositions:   pack.EndPositions,
+				}
+			}
+			d.targetsMu.RUnlock()
+
+			// group messages by vchannels
+			for _, msg := range pack.Msgs {
+				if msg.VChannel() == "" {
+					// for non-dml msg, such as CreateCollection, DropCollection, ...
+					// we need to dispatch it to all the vchannels.
+					for k := range packs {
+						packs[k].Msgs = append(packs[k].Msgs, msg)
+					}
+					continue
+				}
+				if _, ok := packs[msg.VChannel()]; !ok {
+					continue
+				}
+				packs[msg.VChannel()].Msgs = append(packs[msg.VChannel()].Msgs, msg)
+			}
+
+			// dispatches packs to targets
+			lagChannels := make([]string, 0)
+			d.targetsMu.RLock()
+			for vchannel, p := range packs {
+				select {
+				case <-time.After(MaxTolerantLag):
+					log.Warn("time lag is too long for vchannel",
+						zap.String("vchannel", vchannel),
+						zap.Duration("lag", MaxTolerantLag))
+					lagChannels = append(lagChannels, vchannel)
+				case d.targets[vchannel] <- p:
+				}
+			}
+			d.targetsMu.RUnlock()
+
+			// splits lag channels
+			for _, vchannel := range lagChannels {
+				pack.StartPositions[0].ChannelName = vchannel
+				d.lagChan <- pack.StartPositions[0]
+				d.removeTarget(vchannel)
+			}
+			if len(lagChannels) > 0 {
+				log.Warn("dispatcher sent lag signals", zap.Any("lagVchannels", lagChannels))
+			}
+		}
+	}
+}
+
+func (d *dispatcher) setCurPosition(pos *internalpb.MsgPosition) {
+	d.curPosMu.Lock()
+	defer d.curPosMu.Unlock()
+	d.curPos = pos
+}
+
+func (d *dispatcher) getCurPosition() *internalpb.MsgPosition {
```

Comment:
change to getCurTimestmap? Set it to an atomic variable so that the lock can be removed.